### PR TITLE
Explicitly disabled copy/move.

### PR DESCRIPTION
**Pull request by lewie-donckers**
Explicitly disabled copy/move for `Controller` and `TrackingPidLocalPlanner`.

This to avoid accidental copies which could be incorrect/expensive/both.

And to avoid accidental moves which could cause bound callbacks to break.

### CODE REVIEW

File: include/path_tracking_pid/details/no_copy_no_move.hpp [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+{
+public:
+  NoCopyNoMove(const NoCopyNoMove&) = delete;
+  NoCopyNoMove(NoCopyNoMove&&) = delete;
+  NoCopyNoMove& operator=(const NoCopyNoMove&) = delete;
+  NoCopyNoMove& operator=(NoCopyNoMove&&) = delete;
+
+protected:
+  NoCopyNoMove() = default;
+  ~NoCopyNoMove() = default;
```
Comment (Rayman) [inline include/path_tracking_pid/details/no_copy_no_move.hpp:17]: Why does a `NoCopyNoMove` class need a default constructor?

Comment (lewie-donckers) [reply include/path_tracking_pid/details/no_copy_no_move.hpp:17]: It needs a default constructor because it is used as a base class, so it needs to be instantiated (without any arguments).

We need to explicitly default it because we declared (deleted) the copy and move constructors which inhibits the compiler from generating a default constructor itself.

**Comment by Timple** [comment]
> [](https://github.com/lewie-donckers)This to avoid accidental copies which could be incorrect/expensive/both.
And to avoid accidental moves which could cause bound callbacks to break.

This sounds like a generic c++ issue for these kind of constructions.
Isn't there a standard library providing this?

**Comment by lewie-donckers** [comment]
> > This to avoid accidental copies which could be incorrect/expensive/both.
> > And to avoid accidental moves which could cause bound callbacks to break.
> 
> This sounds like a generic c++ issue for these kind of constructions. Isn't there a standard library providing this?

@Timple No, not in the standard library. Boost does have a similar implementation. Since it's so easy to implement and I wasn't sure about the Boost policy, I went for my own.

Want to change to Boost?

**Comment by Timple** [comment]
Preferably. Boost is used a lot by ROS anyway, so I don't mind the extra include.

This construction (and the logic) is a bit hard to follow. And I'm affraid it's also tricky to reproduce (and knowing when) in other packages.

Reducing the code in this PR will increase the chance of that.

File: include/path_tracking_pid/details/no_copy_no_move.hpp (hunk 1 above)
Comment (lewie-donckers) [reply include/path_tracking_pid/details/no_copy_no_move.hpp:17]: The compiler will generate a default constructor but only under certain conditions. And because we declared the copy c'tor and move c'tor (with ` = delete;`), it is not allowed to do so in this case.

I tried to explain that with:

> We need to explicitly default it because we declared (deleted) the copy and move constructors which inhibits the compiler from generating a default constructor itself.

Comment (Rayman) [reply include/path_tracking_pid/details/no_copy_no_move.hpp:17]: In my reply I talked about the *destructor*

Comment (lewie-donckers) [reply include/path_tracking_pid/details/no_copy_no_move.hpp:17]: Oopsie. Indeed the destructor is always available. But we want it to be protected, so we have to default it explicitly.

Base classes should always have a (public) virtual destructor or a protected non-virtual destructor.

Comment (Rayman) [reply include/path_tracking_pid/details/no_copy_no_move.hpp:17]: OK, clear!

**Comment by Rayman** [comment]
Which boost utility should we use then? I could only find the following:
```
BOOST_MOVABLE_BUT_NOT_COPYABLE
BOOST_COPYABLE_AND_MOVABLE
boost::noncopyable
```

**Comment by lewie-donckers** [comment]
> Preferably. Boost is used a lot by ROS anyway, so I don't mind the extra include.
> 
> This construction (and the logic) is a bit hard to follow. And I'm affraid it's also tricky to reproduce (and knowing when) in other packages.
> 
> Reducing the code in this PR will increase the chance of that.

@Timple fixed!

**Review by Rayman — APPROVED** [review]

**Review by Timple — APPROVED** [review]
Thanks!

I do think this is just as readable in the controller source code. But it reduces code maintenance and increases the chance of it being re-used.

**Comment by Timple** [comment]
> And to avoid accidental moves which could cause bound callbacks to break.

This could be added as an in-code comment for future reference. It's an important reason to use this construction I think.
